example-plugins/inj-tyfam-plugin/src/Plugin/InjTyFam.hs

```diff
@@ -0,0 +1,243 @@
+          else skip   -- GHC already handles W-mode
+    ,
+      TcRnTypes.tcPluginStop = \() -> pure ()
+    }
+
+-----
+
+type M = TcPluginM.TcPluginM
+type R = TcRnTypes.TcPluginResult
+
+-- | The no-op
+
+skip :: M R
+skip = pure $ TcRnTypes.TcPluginOk [] []
+
+-- | A @fmap mconcat . sequence@ that short-circuits on
+-- 'TcRnTypes.TcPluginContradiction'
+
+mapMR :: (a -> M R) -> [a] -> M R
+mapMR f = go [] []
+  where
+    go xs ys = \case
+        []   -> pure $ TcRnTypes.TcPluginOk xs ys
+        a:as -> do
+            r <- f a
+            case r of
+              TcRnTypes.TcPluginContradiction{} -> pure r
+              TcRnTypes.TcPluginOk x y          -> go (x <> xs) (y <> ys) as
+
+-----
+
+-- | For a Given constraint @F a1 a2 ... an ~ F b1 b2 ... bn@ where
+-- @F@ is a type family with functional dependencies, we replace the
+-- constraint by new refined constraints.
+--
+-- The new constraints are @ai ~ bi@ for each argument position @i@
+-- that @F@ declares to be injective and where @ai@ and @bi@ are not
+-- manfiestly equal. This addresses @#10833@. Also, if any @ai@ and
+-- @bi@ are manifestly apart, we report a contradiction.
+--
+-- We also include an additional new constraint @F a1 a2 ... an ~ F c1
+-- c2 ... cn@ for which @ci@ is @ai@ if we are also emitting the new
+-- constraint @ai ~ bi@ and @bi@ otherwise.
+--
+-- Note that we cannot discard the @F ... ~ F ...@ constraint, since
+-- some of its arguments might not have fundeps, and so we'd be losing
+-- information. We manually update that constraint in case our new
+-- point-wise equalities are somehow exotic in such a way that
+-- prevents GHC from using them to rewrite the original constraint.
+-- Else we risk emitting new constraints forever.
+--
+-- Note: we harmlessly but wastefully emit @F a1 .. an ~ F a1 .. an@
+-- when @F@ is injective in all of its arguments; GHC will
+-- immedicately discharge that spurious constraint, so this doesn't
+-- lead to divergence.
+
+simplifyG :: FskEnv -> Ct -> M R
+simplifyG fskEnv ct =
+    case predTree of
+      Type.ClassPred{}  -> skip   -- TODO don't need to unpack SCs, right?
+      Type.IrredPred{}  -> skip
+      Type.ForAllPred{} -> skip   -- TODO this is unreachable, right?
+      Type.EqPred eqRel lty rty -> case eqRel of
+          Type.ReprEq -> skip   -- TODO anything useful to do here?
+          Type.NomEq  -> case splitter fskEnv lty rty of
+              Nothing         -> skip
+              Just (tc, args) -> do
+                  mbChanges <- go [] [] args
+                  case mbChanges of
+                    Nothing -> pure $ TcRnTypes.TcPluginContradiction [ct]
+                    Just x  -> ok lty tc x
+  where
+    predTree :: Type.PredTree
+    predTree = Type.classifyPredType (TcRnTypes.ctPred ct)
+
+    -- 'Nothing' means contradiction
+    -- 'Just' means (updated rtys, new constraints)
+    go :: [Xi] -> [Ct] -> [Triple] -> M (Maybe ([Xi], [Ct]))
```

Comment:
`matchInjArgs` (or similar) instead of `go`? Otherwise I think this function is nice and clear (although perhaps acc1 and acc2 could have more self-explanatory names).

Reply:
Done, thanks.